src/parser/mod.rs

```diff
@@ -1,3 +1,89 @@
+use crate::lexer::{Lexer, Token};
+use crate::parser::ast::File;
+use crate::parser::span::Span;
+use bumpalo::Bump;
+use error::{ParseError, ParseResult};
+use std::iter::Peekable;
+
+#[macro_use]
+pub mod tok;
 pub mod ast;
-pub mod parser;
+pub mod error;
+pub mod expr;
+pub mod function;
 pub mod span;
+pub mod statement;
+pub mod toplevel;
+pub mod type_spec;
+
+#[cfg(test)]
+pub mod test;
+
+pub struct Parser<'s, 'a> {
+    _source: &'s str,
+    lexer: Peekable<Lexer<'s>>,
+    arena: &'a Bump,
+}
+
+impl<'s, 'a> Parser<'s, 'a> {
+    pub fn new(source: &'s str, arena: &'a Bump) -> Self {
+        Self {
+            _source: source,
+            lexer: logos::Lexer::new(source).spanned().peekable(),
+            arena,
+        }
+    }
+
+    pub fn parse(&mut self) -> ParseResult<File<'s, 'a>> {
+        self.file()
+    }
```

Comment:
There's some restructuring we could do here, because arguably, the `parse` function should consume the `Parser` and the `arena` should be part of `File`.

Reply:
would you like to do that now?

Reply:
what do you mean with arena should be part of file?

Reply:
oh do you mean that after parse, file gets the ref to arena and parser is gone? That'd make some sense yes

Reply:
Well the `File` is the result of parsing right, so the `Parser` is gonna be dropped, but the arena should stick around. So the `File` (or some type around it) should contain the parsed structure and the arena that all the references live in.

Reply:
though we can also just "forget" the arena here since wer shouldn't allocate more in it anyway. For DST we should build a new arena and afterwards drop this one.